### PR TITLE
support multi var.opts.key

### PR DESCRIPTION
**Pull request by hy05190134**
RT

### CODE REVIEW

File: fw.lua [hunk 1]
```diff
@@ -174,7 +174,7 @@ local function _ac_lookup(self, needle, haystack, ctx)
 	local match, _ac
 
 	-- dictionary creation is expensive, so we use the id of
-	-- the rule as the key to cache the created dictionary 
+	-- the rule as the key to cache the created dictionary
```
Comment (p0pr0ck5) [inline fw.lua]: I appreciate getting rid of an extra space, but I prefer to keep commits atomic and as small as possible

File: fw.lua [hunk 2]
```diff
@@ -233,6 +233,13 @@ local function _parse_collection(self, collection, opts)
 				_collection[n] = value
 			end
 			return _collection
+		end,
+        multi = function(self, collection, value)
```
Comment (p0pr0ck5) [inline fw.lua]: Please fix spacing, your first tab is using space characters, not \t

**Comment by p0pr0ck5** [comment]
Sorry, I was a bit tired when I replied to you last night- there's actually no reason we need to touch `memokey` here. `memokey` is used for caching data transforms, not parsing the data collection. They are unrelated function and `memokey` shouldn't be touched by adding a `multi` `rule.var.opts`.

**Comment by hy05190134** [comment]
adjust for your suggestion

**Comment by p0pr0ck5** [comment]
This looks okay, I will try to do some testing later this week.

File: fw.lua [hunk 3]
```diff
@@ -321,7 +327,7 @@ local function _log_event(self, request_client, request_uri, rule, match)
 				}
 			end
 
-			socket_logger.log(cjson_encode(t) .. "\n")
+			socket_logger.log(cjson.encode(t) .. "\n")
```
Comment (p0pr0ck5) [inline fw.lua]: This is fixed in a separate commit (26a27e6686a4a297a4193ddae5ab2414bc9e408f). Please keep commits atomic.

File: fw.lua [hunk 4]
```diff
@@ -40,7 +40,7 @@ local function _equals(self, a, b)
 		end
 	else
 		_log(self, "Comparing " .. tostring(a) .. " and " .. tostring(b))
-		equals = a == b
+		equals = tostring(a) == tostring(b)
```
Comment (p0pr0ck5) [inline fw.lua]: Please keep commit atomic, this change seems unrelated to the context of this pull request.

File: fw.lua [hunk 5]
```diff
@@ -233,14 +233,20 @@ local function _parse_collection(self, collection, opts)
...
+			return _collection
 		end
 	}
 
-	if (type(collection) ~= "table") then
-		return collection
-	end
-
-	if (opts == nil) then
+	if (type(collection) ~= "table" or opts == nil) then
```
Comment (p0pr0ck5) [inline fw.lua]: Again, please keep commits atomic. This change is unrelated to the context of the pull request.

**Comment by hy05190134** [comment]
I just merge feature/optimise branch into my development branch, is it ok to merge into your development?

**Comment by hy05190134** [comment]
@p0pr0ck5 , it it tested by myself for multi support.
